server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReader.java

```diff
@@ -407,11 +409,15 @@ boolean hasNextPage() throws IOException {
     }
 
     // make sure firstPageReader won't be null while cachedPageReaders has more cached page readers
-    while (firstPageReader == null && !cachedPageReaders.isEmpty()) {
-      firstPageReader = cachedPageReaders.poll();
-      if (!cachedPageReaders.isEmpty()
-          && orderUtils.isOverlapped(firstPageReader.getStatistics(),
-          cachedPageReaders.peek().getStatistics())) {
+    while (firstPageReader == null && (!seqPageReaders.isEmpty() || !unSeqPageReaders.isEmpty())) {
+
+      initFirstPageReader();
+
+      if ((!seqPageReaders.isEmpty() && orderUtils
+          .isOverlapped(firstPageReader.getStatistics(), seqPageReaders.get(0).getStatistics()))
+          || (!unSeqPageReaders.isEmpty() && orderUtils
+          .isOverlapped(firstPageReader.getStatistics(),
+              unSeqPageReaders.peek().getStatistics()))) {
```

Comment:
better to extract these complicate condition statements into a private method, since they appear not only once in this file (for example, in lines 395-398)

Reply:
You are right, fixed~